python/tests/io/test_base.py

```diff
@@ -0,0 +1,188 @@
+
+        # Test opening and reading the file
+        f = input_file.open()
+        data = f.read()
+        assert data == b"foo"
+
+
+@pytest.mark.parametrize("CustomOutputFile", [LocalOutputFile])
+def test_custom_local_output_file(CustomOutputFile):
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        output_file_location = os.path.join(tmpdirname, "foo.txt")
+
+        # Instantiate an output file
+        output_file = CustomOutputFile(location=f"file://{output_file_location}")
+
+        # Create the output file and write to it
+        f = output_file.create()
+        f.write(b"foo")
+
+        # Confirm that bytes were written
+        with open(output_file_location, "rb") as f:
+            assert f.read() == b"foo"
+
+
+@pytest.mark.parametrize("CustomOutputFile", [LocalOutputFile])
+def test_custom_local_output_file_with_overwrite(CustomOutputFile):
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        output_file_location = os.path.join(tmpdirname, "foo.txt")
+
+        # Create a file in the temporary directory
+        with open(output_file_location, "wb") as f:
+            f.write(b"foo")
+
+        # Instantiate an output file
+        output_file = CustomOutputFile(location=f"file://{output_file_location}")
+
+        # Confirm that a FileExistsError is raised when overwrite=False
+        with pytest.raises(FileExistsError):
+            f = output_file.create(overwrite=False)
+            f.write(b"foo")
+
+        # Confirm that the file is overwritten with overwrite=True
+        f = output_file.create(overwrite=True)
+        f.write(b"bar")
+        with open(output_file_location, "rb") as f:
+            assert f.read() == b"bar"
+
+
+@pytest.mark.parametrize("CustomOutputFile", [LocalOutputFile])
+def test_output_file_to_input_file(CustomOutputFile):
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        output_file_location = os.path.join(tmpdirname, "foo.txt")
+
+        # Create an output file instance
+        output_file = CustomOutputFile(location=f"file://{output_file_location}")
+
+        # Create the output file and write to it
+        f = output_file.create()
+        f.write(b"foo")
+
+        # Convert to an input file and confirm the contents
+        input_file = output_file.to_input_file()
+        with input_file.open() as f:
+            assert f.read() == b"foo"
+
+
+@pytest.mark.parametrize(
+    "CustomFileIO,CustomInputFile,CustomOutputFile",
+    [(LocalFileIO, LocalInputFile, LocalOutputFile)],
+)
+def test_custom_file_io(CustomFileIO, CustomInputFile, CustomOutputFile):
+    # Instantiate the file-io and create a new input and output file
+    file_io = CustomFileIO()
+    input_file = file_io.new_input(location="file://foo")
+    output_file = file_io.new_output(location="file://bar")
+
+    assert isinstance(input_file, CustomInputFile)
+    assert isinstance(output_file, CustomOutputFile)
```

Comment:
I don't think that we need to test the types that are produced by FileIO, other than to ensure that they are InputFile and OutputFile instances. But we should get that with type checking for free without a test.

Reply:
Sounds good, removed this test.